### PR TITLE
[1429] A failed CONNACK should be sent, when a CONNECT packet exceed the max packet size version 2

### PR DESCRIPTION
**Pull request by tgracchus**
**Motivation**
EXPECTED BEHAVIOR
A mqtt5 client get a negative connack with code 149; The CONNECT packet exceeded the maximum permissible size.
 

ACTUAL BEHAVIOR
Currently if a CONNECT packet exceeds the configured max packet size we just close the connection.

This PR fixes the issue and makes hivemq behaves as expected

Resolves #<issue>

**Changes*

PR
master https://github.com/hivemq/hivemq-enterprise/pull/2051/files 
4.9: https://github.com/hivemq/hivemq-enterprise/pull/2052 
CE: https://github.com/hivemq/hivemq-community-edition/pull/423
CE Test:[ https://github.com/hivemq/hivemq-community-edition-test/pull/71](https://github.com/hivemq/hivemq-community-edition-test/pull/71)

Running IT test is ok!
https://github.com/hivemq/hivemq-community-edition-composite/pull/35

### CODE REVIEW

**Comment by sauroter** [comment]
[clabot:check]

**Comment by sauroter** [comment]
@cla-bot check

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 1]
```diff
@@ -174,4 +176,149 @@ public void test_second_connect_received() {
         assertFalse(channel.isOpen());
 
     }
+
+    @Test
+    public void test_connect_mqtt5_packet_size_too_large() {
+        final byte[] mqtt5_connect = {
```
Comment (mchernyakov) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:182]: please make variable-names in camelcase (182,208,234)

Comment (tgracchus) [reply src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java]: changed

File: src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java [hunk 2]
```diff
@@ -113,127 +120,192 @@ protected void decode(
...
+                "Sent CONNECT exceeded the maximum permissible size",
+                Mqtt5ConnAckReasonCode.PACKET_TOO_LARGE,
+                ReasonStrings.CONNACK_PACKET_TOO_LARGE);
+    }
+
+    private static boolean isAlreadyConnected(ClientConnectionContext clientConnectionContext) {
+        return clientConnectionContext.getProtocolVersion() != null;
+    }
+
+    private Message handledMessage(
```
Comment (mchernyakov) [inline src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java:210]:
```suggestion
    private @Nullable Message handledMessage(
```

File: src/main/java/com/hivemq/codec/decoder/MqttConnectDecoder.java [hunk 3]
```diff
@@ -117,6 +114,19 @@ public MqttConnectDecoder(
         clientConnectionContext.setProtocolVersion(protocolVersion);
         clientConnectionContext.setConnectReceivedTimestamp(System.currentTimeMillis());
 
+        return protocolVersion;
+    }
+
```
Comment (mchernyakov) [inline src/main/java/com/hivemq/codec/decoder/MqttConnectDecoder.java:119]:
```suggestion

```

File: src/main/java/com/hivemq/codec/decoder/MqttConnectDecoder.java [hunk 4]
```diff
@@ -126,6 +136,7 @@ public MqttConnectDecoder(
         }
     }
 
+
```
Comment (mchernyakov) [inline src/main/java/com/hivemq/codec/decoder/MqttConnectDecoder.java:139]:
```suggestion

```

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 5]
```diff
@@ -174,4 +176,149 @@ public void test_second_connect_received() {
...
+                0, 0,
+                //   properties
+                0,
+                // payload
+                //   client identifier
+                0, 4, 't', 'e', 's', 't'};
+
+        test_connect_packet_size_too_large(mqtt31Connect);
+    }
+
```
Comment (mchernyakov) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:257]:
```suggestion

```

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 6]
```diff
@@ -174,4 +176,149 @@ public void test_second_connect_received() {
...
+    public void test_publish_mqtt3_1_1_packet_size_too_large() {
+        test_publish_packet_size_too_large(ProtocolVersion.MQTTv3_1_1);
+    }
+
+    @Test
+    public void test_publish_mqtt3_1_packet_size_too_large() {
+        test_publish_packet_size_too_large(ProtocolVersion.MQTTv3_1);
+    }
+
+    private void test_publish_packet_size_too_large(final @NotNull ProtocolVersion protocolVersion) {
```
Comment (mchernyakov) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:274]: that's a method, you can name it in normal java-way.
```suggestion
    private void testPublishPacketSizeTooLarge(final @NotNull ProtocolVersion protocolVersion) {
```

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 7]
```diff
@@ -174,4 +176,149 @@ public void test_second_connect_received() {
...
+        final ByteBuf buf = Unpooled.buffer();
+        buf.writeBytes(publish);
+        channel.writeInbound(buf);
+
+        //verify that the client was disconnected
+        assertFalse(channel.isOpen());
+    }
+
+
+    private void test_connect_packet_size_too_large(byte[] connect) {
```
Comment (mchernyakov) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:308]:
```suggestion
    private void testConnectPacketSizeTooLarge(byte[] connect) {
```

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java (hunk 6 above)
Comment (tgracchus) [reply src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:273]: test have the same name than in enterprise, those are legacy test, I rather keep them as they are just to be similar to the old ones

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 8]
```diff
@@ -174,4 +176,148 @@ public void test_second_connect_received() {
...
+        };
+
+        final ByteBuf buf = Unpooled.buffer();
+        buf.writeBytes(publish);
+        channel.writeInbound(buf);
+
+        //verify that the client was disconnected
+        assertFalse(channel.isOpen());
+    }
+
```
Comment (mchernyakov) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:305]:
```suggestion

```

**Review by mchernyakov — APPROVED** [review]

File: src/main/java/com/hivemq/bootstrap/netty/ChannelDependencies.java [hunk 9]
```diff
@@ -216,6 +220,10 @@ public MqttConnectDecoder getMqttConnectDecoder() {
         return mqttConnectDecoder;
     }
 
+    public MqttConnacker getMqttConnacker() {
```
Comment (Remit) [inline src/main/java/com/hivemq/bootstrap/netty/ChannelDependencies.java:223]:
```suggestion
    public @NotNull MqttConnacker getMqttConnacker() {
```

File: src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java [hunk 10]
```diff
@@ -128,112 +215,78 @@ protected void decode(
...
 
-        if (protocolVersion != null && messageType == CONNECT) {
-            mqttServerDisconnector.logAndClose(clientConnectionContext.getChannel(),
-                    "A client (IP: {}) sent second CONNECT message. This is not allowed. Disconnecting client.",
-                    "Sent second CONNECT message");
-            buf.clear();
-            return;
-        }
-
-        globalMQTTMessageCounter.countInboundTraffic(readableBytes);
```
Comment (Remit) [inline src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java:159]: I think this metric update got lost

File: src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java [hunk 11]
```diff
@@ -128,112 +215,78 @@ protected void decode(
...
+                    return null;
             }
         }
-
-        if (message == null) {
-            buf.clear();
-            return;
-        }
-
-        globalMQTTMessageCounter.countInbound(message);
```
Comment (Remit) [inline src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java:235]: I think this metric update got lost

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 12]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
         assertFalse(channel.isOpen());
 
     }
+
+    @Test
+    public void test_connect_mqtt5_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:181]: please name new tests according to the conventions

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 13]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
...
+                0,
+                // payload
+                //   client identifier
+                0, 4, 't', 'e', 's', 't'};
+
+        test_connect_packet_size_too_large(mqtt5Connect);
+    }
+
+    @Test
+    public void test_connect_mqtt3_1_1_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:207]: please name new tests according to the conventions

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 14]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
...
+                0,
+                // payload
+                //   client identifier
+                0, 4, 't', 'e', 's', 't'};
+
+        test_connect_packet_size_too_large(mqtt311Connect);
+    }
+
+    @Test
+    public void test_connect_mqtt3_1_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:233]: please name new tests according to the conventions

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 15]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
...
+                0,
+                // payload
+                //   client identifier
+                0, 4, 't', 'e', 's', 't'};
+
+        test_connect_packet_size_too_large(mqtt31Connect);
+    }
+
+    @Test
+    public void test_publish_mqtt5_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:259]: please name new tests according to the conventions

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 16]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
...
+        test_publish_packet_size_too_large(ProtocolVersion.MQTTv5);
+    }
+
+    @Test
+    public void test_publish_mqtt3_1_1_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:264]: please name new tests according to the conventions

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java [hunk 17]
```diff
@@ -174,4 +176,147 @@ public void test_second_connect_received() {
...
+        test_publish_packet_size_too_large(ProtocolVersion.MQTTv3_1_1);
+    }
+
+    @Test
+    public void test_publish_mqtt3_1_packet_size_too_large() {
```
Comment (Remit) [inline src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:269]: please name new tests according to the conventions

**Review by Remit — CHANGES_REQUESTED** [review]
Please add metrics updates that got removed

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java (hunk 12 above)
Comment (tgracchus) [reply src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:181]: this class is legacy, I'm just following the naming convection it already had.
See https://github.com/hivemq/hivemq-community-edition/pull/423/files/c7d4ce9e81b853c7dc35a6c09c7f795925f1016d#diff-11e37a0c7b609495e61a553c8ab4b5e851b84af733b30c4a26f70c05bede76b4L138

Should we change all the test names here?

File: src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java (hunk 13 above)
Comment (tgracchus) [reply src/test/java/com/hivemq/codec/decoder/MQTTMessageDecoderTest.java:207]: its there, for some reason github its not updateing the repo files

File: src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java (hunk 11 above)
Comment (tgracchus) [reply src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java:235]: true, added

File: src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java (hunk 10 above)
Comment (tgracchus) [reply src/main/java/com/hivemq/codec/decoder/MQTTMessageDecoder.java:159]: added back

**Review by Remit — APPROVED** [review]

**Review by Remit — APPROVED** [review]

**Pull request merged**